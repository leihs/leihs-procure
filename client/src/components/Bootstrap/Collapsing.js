import React from 'react'
import PropTypes from 'prop-types'
import f from 'lodash'
import Icon from '../Icons'

export class Collapsing extends React.Component {
  static defaultProps = { startOpen: false, canToggle: true }
  static propTypes = { id: PropTypes.string.isRequired }

  constructor(props) {
    super()
    this.state = { isOpen: props.canToggle === false || props.startOpen }
  }

  onToggleOpen(event) {
    event.preventDefault()
    if (!this.props.canToggle) return
    this.setState(
      s => ({ isOpen: !s.isOpen }),
      () => this.handleCallback()
    )
  }

  handleCallback() {
    if (f.isFunction(this.props.onChange)) {
      this.props.onChange(this.state)
    }
  }

  render({ props: { id, children, canToggle }, state: { isOpen } } = this) {
    const toggleOpen = e => this.onToggleOpen(e)

    return children({
      canToggle,
      toggleOpen,
      isOpen,
      Caret: isOpen ? Icon.CaretDown : Icon.CaretRight,
      collapsedProps: {
        id: `${id}-content`,
        'aria-labelledby': id ? `${id}-toggle` : false
      },
      togglerProps: {
        onClick: toggleOpen,
        id: `${id}-toggle`,
        'aria-expanded': isOpen ? 'true' : 'false',
        'aria-controls': `${id}-content`
      }
    })
  }
}
