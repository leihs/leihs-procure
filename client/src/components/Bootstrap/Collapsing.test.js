import renderer from 'react-test-renderer'
import { examples } from './Collapsing.examples'
import { it, expect } from 'vitest'

examples.forEach(({ name, content }) => {
  it(`${name}: renders correctly`, () => {
    expect(renderer.create(content).toJSON()).toMatchSnapshot()
  })
})
