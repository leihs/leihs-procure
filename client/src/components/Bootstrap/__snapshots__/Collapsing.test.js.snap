// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`undefined: renders correctly 1`] = `
<div
  className="row"
>
  <div
    className="col-sm"
  >
    <div>
      <div
        aria-controls="collapse-example-basic-content"
        aria-expanded="true"
        id="collapse-example-basic-toggle"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-caret-down fa-w-10 fa-fw mr-1"
          data-icon="caret-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={{}}
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
            fill="currentColor"
            style={{}}
          />
        </svg>
        Collapsed Title
      </div>
      <div
        aria-labelledby="collapse-example-basic-toggle"
        id="collapse-example-basic-content"
      >
        Collapsed Content!
      </div>
    </div>
  </div>
  <div
    className="col-sm"
  >
    <pre>
      <code>
        
&lt;Collapsing id="collapse-example-1" startOpen&gt;
  {({ isOpen, togglerProps, collapsedProps, Caret }) =&gt; (
    &lt;div&gt;
      &lt;div {...togglerProps}&gt;
        &lt;Caret spaced /&gt;
        Collapsed Title
      &lt;/div&gt;
      {isOpen && &lt;div {...collapsedProps}&gt;Collapsed Content!&lt;/div&gt;}
    &lt;/div&gt;
  )}
&lt;/Collapsing&gt;
        
      </code>
    </pre>
  </div>
</div>
`;

exports[`undefined: renders correctly 2`] = `
<div
  className="row"
>
  <div
    className="col-sm"
  >
    <div
      className="card mb-3"
    >
      <div
        aria-controls="collapse-example-card-content"
        aria-expanded="true"
        className="card-header cursor-pointer pl-2"
        id="collapse-example-card-toggle"
        onClick={[Function]}
      >
        <h1
          className="mb-0 h4 d-inline-block"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-caret-down fa-w-10 fa-fw mr-1"
            data-icon="caret-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={{}}
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
              fill="currentColor"
              style={{}}
            />
          </svg>
          Collapsed Title
        </h1>
      </div>
      <div
        aria-labelledby="collapse-example-card-toggle"
        className="card-body pl-2"
        id="collapse-example-card-content"
      >
        <p
          className="h5 p-4"
        >
          Collapsed Content!
        </p>
      </div>
    </div>
  </div>
</div>
`;
